### ISSUE TITLE
Buildserver documentation

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#1974]
Documentation for build server, please describe:

- [x] tutorial for Jenkinsfile for people not knowing Jenkins, Groovy, or Dockerfiles; with hints/references to:
  - [x] how to add a docker-based build agent
  - [x] how to configure the build job in Jenkins (if Jenkins is newly installed)
  - [x] how to identify which docker image should be used in issue reports (so how to write a good issue which allows reproducibility with our docker images)
  - [x] how to get a docker image to reproduce a problem (for devs)
  - [x] how docker containers and Jenkinsfile are connected.
  - [x] how to add a build job (which compiles Elektra)
  - [x] how to add a script to be executed on the build server (which does checks)
- [x] some more comments in Jenkinsfile (about structure, main, ...)
- [x] when which job gets triggered (push to master, PRs, or debian, see also #2027)
- [x] section about deployment in BUILDSERVER.md, see also #2003
- [x] describe which Jenkinsfile is used (if PR and master both make changes to the Jenkinsfile)
- [x] bring scripts/docker/README.md up-to-date
- [x] overview of Dockerfiles in scripts/docker

If some parts of the documentation already exist for this release, it would be great.

**Comment by markus2330** [comment]
A question: Why is deleteDir() after the build actions? (As seen in #2000) This means there are never files there for inspection? Or isn't this deleteDir() executed in these cases?

Having some documentation how to access files after a failed build (in particular the Testing.xml for memleaks is often useful) is important.

**Comment by ingwinlu** [comment]
workspaces should be treated as ephemeral. The new build job should handle cleanup before and after builds automatically (it is just there because it was needed before).

If you want to archive files you should use `archive` which is a wrapper around `archiveArtifacts`. Then you can download them from the jenkins master (https://build.libelektra.org/jenkins/job/libelektra/job/master/7/artifact/artifacts/).

**Comment by markus2330** [comment]
Is the information in scripts/docker/README.md still correct? @petermax2 asked about reproducing #2035 on command-line.

**Comment by ingwinlu** [comment]
> Is the information in scripts/docker/README.md still correct?

Probably. I never touched the Dockerfile that it describes.

> reproducing #2035 on command-line

there are some problems with that:
* the original problem with the error that occured during key import never showed up when I ran every step manually
* the readme describes using a docker image that is not used for the unit tests

**Comment by markus2330** [comment]
Yes, we should update the README that there are now several Dockerfiles. And the most important Dockerfiles for devs are now the one written by you. It would be great to have a description how to locally reproduce what the build server does.

**Comment by ingwinlu** [comment]
> how to identify which docker image should be used in issue reports (so how to write a good issue which allows reproducibility with our docker images)

Difficult. Depends on how the issue occurred and can only be answered depending on the input.
For example if a report with clang comes up you would go for the sid image (as this is where we run our clang tests). For pretty much anything else it is the current debian stable image.

**Comment by markus2330** [comment]
Maybe you can simply print which Dockerfile is used at the start of the build jobs?

Would be great to have the most important information (Dockerfile, git commit sha1, which subtask, URL to buildjob, ...) at one spot, so that everything can be copied and pasted when reporting an issue.

And even better would be that the build server directly posts this information into the PRs when build jobs fail. (Does not work on failing master, though.)

**Comment by ingwinlu** [comment]
> Maybe you can simply print which Dockerfile is used at the start of the build jobs?

that is already happening

> Would be great to have the most important information (Dockerfile, git commit sha1, which subtask, URL to buildjob, ...) at one spot, so that everything can be copied and pasted when reporting an issue.

huh? are you mixing some things into one here?
* For open PR's all this information is available in the build job that can be referenced via url
* for everything else it is of little value as:
  * the dockerfile might have been modified
  * the sources in the git repository could have been modified
  * the rest does not exist for reporting issues

> And even better would be that the build server directly posts this information into the PRs when build jobs fail. (Does not work on failing master, though.)

that would be super spammy. the information is automatically made available for each commit the build jobs run.

**Comment by markus2330** [comment]
> that is already happening

Then please document where it is, I could not find it. With the name of the Dockerfile devs should already know how to reproduce something locally. Or am I missing something?

>  For open PR's all this information is available in the build job that can be referenced via url

Again, I do not see where.  Especially the output, of which git commit is build, is quite hard to find. (The output is too verbose.)

> are you mixing some things into one here?

Yes but on purpose: currently relevant information is quite distributed. Nobody will copy&paste the name of the Dockerfile, the node name, the sha1 if it is hidden in 3 different parts.

> that would be super spammy.

It might be but it seems to be standard for many other initiatives because the build job status is something that can be easily overlooked. It also does not trigger notifications. And the build job status gives only a link but nothing about the failure.

**Comment by ingwinlu** [comment]
> Would be great to have the most important information (Dockerfile, git commit sha1, which subtask, URL to buildjob, ...) at one spot, so that everything can be copied and pasted when reporting an issue.

**dockerfile**: I admit it is a bit hidden:
![docker image](https://user-images.githubusercontent.com/4435962/41042732-ab06afee-69a2-11e8-88bc-61db8ffd9880.png)

It was sufficient for me until now. It should be trivial for anybody to add the image.id as an additional printed message that is not as hidden.

**commit**: 
![commit](https://user-images.githubusercontent.com/4435962/41042800-d939a74a-69a2-11e8-9438-48cd425685e1.PNG)
You can also click `CHANGES` to see all commits since the last builds (clickable with commit text).
As for the output of checkout scm: It already echos what it is trying to do (lines without >) and the actual git commands (lines with >).

**subtask**:
![taskname](https://user-images.githubusercontent.com/4435962/41042980-3fbfe8f8-69a3-11e8-9af4-200115b4c00e.PNG)

Here I feel you are not entirely serious. The subtask name is listed in the pipeline graphic, the pipeline log start. And once more during the debug output where it also lists on what machine it ran.

> comments

We currently have no plugin activated that provides that functionality. We discussed this already. Especially with how long it took to fix the issues in the github source plugin (mind you a plugin maintained by the jenkins core team) I would only add plugins that are absolutely needed (and once everything else is in place).

**Comment by markus2330** [comment]
@ingwinlu I added two more documentation tasks:

- [ ] how to add a docker-based build agent
- [ ] how to configure the build job in Jenkins (if Jenkins is newly installed)

because I plan to add an i7 system (see #2020) and you never know when we need to reinstall/reconfigure Jenkins from scratch.

Is there progress with dockerization of the debian-stable-mm system? This is a good chance to describe what is needed to add a docker-based build agent.

Btw. why does it say that the last success of "libelektra" is 14 days ago? (On the front page https://build.libelektra.org/jenkins/)

**Comment by ingwinlu** [comment]
@markus2330 good points :+1: 

> Is there progress with dockerization of the debian-stable-mm system?

No progress. I stopped after the update killed the unstable VM.

> Btw. why does it say that the last success of "libelektra" is 14 days ago?

Because that was the last time the indexation ran for the 'libelektra' folder. Keep in mind that this is not the 'libelektra/master' build job or any of the PR's.
I looked around if one could override the status to the one of master but that is currently not a feature.

**Comment by ingwinlu** [comment]
> Seems like I do not understand what the "libelektra" build job is. So it is a folder that contains build jobs? 

Yes. Though this one is expanded for Github functionaility. It also takes care of setting properties for all jobs nested inside the folder or creating all the build jobs it receives info via webhooks or indexing.

> How to rerun indexation?

You can trigger it from the folder in the old view, in the actions list (the thing on the left above the build queue). You should never have to though as it is run automatically if it has not received any events for x amount of time.

> some docu

I would not know where to start. The basics are covered by the Jenkins documentation. In depth is completely out of scope for the Elektra documentation.

You really do not need to know anything about it to use it though.

**Comment by markus2330** [comment]
I updated the top post. Unfortunately, a lot of information about the build server is only available in the GitHub discussions we had. It would be great if this information could be put into our docu.
